src/Simd/QuaternionDouble.cs

```diff
@@ -0,0 +1,157 @@
+//
+// QuaternionDouble.cs:
+//     This represents the native quaternion double type, which is 32 bytes.
+//
+//
+// Authors:
+//     Stephane Delcroix <stephane@delcroix.org>
+//
+// Copyright (c) 2021 Microsoft Inc
+//
+
+using System;
+using System.Runtime.InteropServices;
+
+#if NET
+namespace CoreGraphics
+{
+	[StructLayout (LayoutKind.Sequential)]
+	public struct NQuaterniond : IEquatable<NQuaterniond>
+	{
+		double x;
+		double y;
+		double z;
+		double w;
+
+		public NQuaterniond (double x, double y, double z, double w)
+		{
+			this.x = x;
+			this.y = y;
+			this.z = z;
+			this.w = w;
+		}
+
+		public NQuaterniond (NVector3d xyz, double w)
+		{
+			x = xyz.X;
+			y = xyz.Y;
+			z = xyz.Z;
+			this.w = w;
+		}
+
+		public NVector3d Xyz
+		{
+		    get => new NVector3d(x, y, z);
+		    set
+		    {
+			x = value.X;
+			y = value.Y;
+			z = value.Z;       
+		    }
+		}
+
+		public double X
+		{
+		    get => x;
+		    set => x = value;
+		}
+
+		public double Y
+		{
+		    get => y;
+		    set => y = value;
+		}
+
+		public double Z
+		{
+		    get => z;
+		    set => z = value;
+		}
+
+		public double W
+		{
+		    get => w;
+		    set => w = value;
+		}
+
+		public static bool operator == (NQuaterniond left, NQuaterniond right)
+		{
+			return left.Equals (right);
+		}
+
+		public static bool operator != (NQuaterniond left, NQuaterniond right)
+		{
+			return !left.Equals (right);
+		}
+
+		public override string ToString ()
+		{
+			return $"({X}, {Y}, {Z}, {W})";
+		}
+
+		public override int GetHashCode ()
+		{
+			return X.GetHashCode () ^ Y.GetHashCode () ^ Z.GetHashCode () ^ W.GetHashCode ();
+		}
+
+		public override bool Equals (object obj)
+		{
+			if (!(obj is NQuaterniond))
+				return false;
+
+			return Equals ((NQuaterniond) obj);
+		}
+
+		public bool Equals (NQuaterniond other)
+		{
+			return X == other.X && Y == other.Y && Z == other.Z && W == other.W;
+		}
+
+		internal double Length
+		{
+			get
+			{
+				return (double)System.Math.Sqrt(W * W + Xyz.LengthSquared);
+			}
+		}
+        
+		internal void Normalize()
+		{
+			double scale = 1.0f / this.Length;
+			Xyz *= scale;
+			W *= scale;
+		}
+
+		internal void ToAxisAngle(out NVector3d axis, out double angle)
+		{
+		    NVector4d result = ToAxisAngle();
```

Comment:
Code formatting isn't following our style guide, the entire file should be re-formatted.